### PR TITLE
supports collective communicated training

### LINKED ISSUE #18174: add 'collective op'
add collective operations  to support collective data exchange ,support for  allreduce /allgather /broadcast /slicegather .

### PR DESCRIPTION
**Pull request by LLMHao**
fix https://github.com/PaddlePaddle/Paddle/issues/18174

### CODE REVIEW

File: paddle/fluid/operators/collective/c_allgather_op.cc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+namespace operators {
+
+class CAllGatherOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should not be null");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"),
+                   "Output(Out) of SyncFCGather op should not be null.");
+    int nranks = ctx->Attrs().Get<int>("nranks");
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_allgather_op.cc:30]: Do you have check on nranks?

File: paddle/fluid/operators/collective/c_allgather_op.cc [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        .SetDefault(0);
+    AddAttr<bool>(
+        "use_calc_stream",
+        "(bool default false) eject CUDA operations to calculation stream.")
+        .SetDefault(false);
+    AddAttr<int>("nranks",
+                 "Total trainer count of the distributed training job");
+    AddComment(R"DOC(
+***CAllGather Operator***
+
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_allgather_op.cc:54]: All Gather should be clarified more clearly, you can also link a url to nvidia's API.

File: paddle/fluid/operators/collective/c_allgather_op.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    auto place = ctx.GetPlace();
+    PADDLE_ENFORCE(is_gpu_place(place),
+                   "CAllGatherOp can run on gpu place only for now.");
+#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+    auto in = ctx.Input<framework::Tensor>("X");
+    auto out = ctx.Output<framework::Tensor>("Out");
+    ncclDataType_t dtype = platform::ToNCCLDataType(in->type());
+
+    int rid = ctx.Attr<int>("ring_id");
+    auto comm = platform::NCCLCommContext::Instance().Get(rid);
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_allgather_op.h:46]: do you need to check whether rid exists?

File: paddle/fluid/operators/collective/c_allreduce_op.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    auto out = ctx.Output<framework::Tensor>("Out");
+
+    ncclDataType_t dtype = platform::ToNCCLDataType(in->type());
+    int64_t numel = in->numel();
+    const void* sendbuff = in->data<void>();
+    out->Resize(in->dims());
+    void* recvbuff = out->mutable_data<T>(place);
+
+    int rid = ctx.Attr<int>("ring_id");
+    auto comm = platform::NCCLCommContext::Instance().Get(rid);
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_allreduce_op.h:50]: same as above

File: paddle/fluid/operators/collective/c_broadcast_op.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    cudaStream_t stream = nullptr;
+    if (ctx.Attr<bool>("use_calc_stream")) {
+      auto dev_ctx = platform::DeviceContextPool::Instance().Get(place);
+      stream = static_cast<platform::CUDADeviceContext*>(dev_ctx)->stream();
+    } else {
+      stream = comm->stream();
+    }
+
+    int root = ctx.Attr<int>("root");
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_broadcast_op.h:57]: do you need to check the range of root?

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+               const framework::AttributeMap& attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+
+  void RunImpl(const framework::Scope& scope,
+               const platform::Place& dev_place) const override {
+    platform::DeviceContextPool& pool = platform::DeviceContextPool::Instance();
+    // put nccl id in CPUPlace
+    auto& dev_ctx = *pool.Get(platform::CPUPlace());
+    int rank = Attr<int>("rank");
+    framework::Scope& local_scope = scope.NewScope();
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_gen_nccl_id_op.cc:47]: local scope can be deleted, is it ok to NewScope here?

File: paddle/fluid/operators/collective/c_allgather_op.cc (hunk 2 above)
Comment (LLMHao) [reply paddle/fluid/operators/collective/c_allgather_op.cc:54]: Done

File: paddle/fluid/operators/collective/c_allgather_op.cc (hunk 1 above)
Comment (LLMHao) [reply paddle/fluid/operators/collective/c_allgather_op.cc:30]: done.add check nranks>=2

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc (hunk 6 above)
Comment (gavin1332) [reply paddle/fluid/operators/collective/c_gen_nccl_id_op.cc:47]: nccl_id is an output var which is persistable, and this local scope is created for nccl_id comminication.

File: paddle/fluid/operators/collective/c_broadcast_op.h (hunk 5 above)
Comment (LLMHao) [reply paddle/fluid/operators/collective/c_broadcast_op.h:57]: Done. root should be [0,nranks).

File: python/paddle/fluid/incubate/fleet/base/role_maker.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.uto-merging paddle/fluid/operators/shard_index_op.h
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/base/role_maker.py]: typo??

Comment (LLMHao) [reply python/paddle/fluid/incubate/fleet/base/role_maker.py]: done

File: paddle/fluid/operators/collective/c_comm_init_op.cc [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+              const framework::VariableNameMap& outputs,
+              const framework::AttributeMap& attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+
+  void RunImpl(const framework::Scope& scope,
+               const platform::Place& place) const override {
+    PADDLE_ENFORCE(is_gpu_place(place),
+                   "CCommInitOp can run on gpu place only.");
+
+    auto var = scope.FindVar(Input("X"));
```
Comment (sneaxiy) [inline paddle/fluid/operators/collective/c_comm_init_op.cc:46]: `X` should be output?

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
```
Comment (sneaxiy) [inline paddle/fluid/operators/collective/c_gen_nccl_id_op.cc:47]: Seems that we only need `NewScope` when `rank != 0`, and this temporarily Scope can be deleted immediately after `GetIdByServer` is called?

File: paddle/fluid/operators/collective/c_sync_calc_stream_op.cc [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    }
+#else
+    PADDLE_THROW("PaddlePaddle should compile with GPU.");
+#endif
+  }
+};
+
+class CSyncCalcStreamOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() {
```
Comment (sneaxiy) [inline paddle/fluid/operators/collective/c_sync_calc_stream_op.cc:59]: `c_sync_calc_stream` op has no in/out, which may cause graph dependency analysis failure in the future...

File: paddle/fluid/operators/one_hot_op.cc [hunk 11]
```diff
@@ -86,6 +86,11 @@ class OneHotOpMaker : public framework::OpProtoAndCheckerMaker {
                  "An integer to specify the data type of one-hot "
                  "vector. The default value is FP32.")
         .SetDefault(paddle::framework::proto::VarType::FP32);
+    AddAttr<bool>("allow_out_of_range",
```
Comment (sneaxiy) [inline paddle/fluid/operators/one_hot_op.cc:89]: Why add this? Is it better to report error when out of bound occurs?

File: python/paddle/fluid/framework.py [hunk 12]
```diff
@@ -2756,6 +2758,9 @@ def __init__(self):
         # the distributed lookup table names
         self._distributed_lookup_table = None
 
+        # the original program before collective transpiling
+        self.origin_program = None
```
Comment (sneaxiy) [inline python/paddle/fluid/framework.py]: Do we really need to store the `origin_program` in such a way? It seems too fancy and hard-code.

File: paddle/fluid/operators/collective/c_comm_init_op.cc (hunk 8 above)
Comment (renhaosw) [reply paddle/fluid/operators/collective/c_comm_init_op.cc]: 'X': nccl_id_var  is created by op c_gen_nccl_id as output which called before c_comm_init ,here c_comm_init op use it as input to create nccl communicator

File: paddle/fluid/operators/one_hot_op.cc (hunk 11 above)
Comment (gavin1332) [reply paddle/fluid/operators/one_hot_op.cc:89]: we added distributed fc support which separates the fc parameter to different gpu cards, and it collaborates with labels to calculate loss. In our case, the label should firstly recalculated by trainer numbers, and then change to one hot representation, e.g.
two trainer: t1, t2, with data parallel
class dimension: 10, and each trainer takes 5, i.e. we divided the whole class dimension into 2.
batched label in t1: 2, 4, 6, 8
batched label in t2: 2, 4, 6, 8 (the same with t1 for simple illustration)
firstly, label is recalculated to
t1: 2, 4, -1, -1
t2: -1, -1, 1, 3, in which, (6-5)%5=1, (8-5)%5=3
then we get one hot representation as
t1: 00100,00001,00000,00000
t2: 00000,00000,01000,00100
u see, we need correctly convert `out_of_range` label -1 to uniform 0 representation, that why we add this api

File: python/paddle/fluid/framework.py (hunk 12 above)
Comment (gavin1332) [reply python/paddle/fluid/framework.py]: we really need suggestion! we are explore a new distributed algorithm named LocalSGD, which do local update with original non-transpiled program for multiple steps, and then update with the transpiled program.

Comment (gavin1332) [reply python/paddle/fluid/framework.py]: I found that it is better to store the original program in our training code but not in the framework!
we will fix it. Thanks!

File: paddle/fluid/operators/one_hot_op.cc (hunk 11 above)
Comment (sneaxiy) [reply paddle/fluid/operators/one_hot_op.cc:89]: I see. Thanks!

File: paddle/fluid/operators/collective/c_comm_init_op.cc (hunk 8 above)
Comment (LLMHao) [reply paddle/fluid/operators/collective/c_comm_init_op.cc:46]: it used in start_up_program.

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc (hunk 9 above)
Comment (LLMHao) [reply paddle/fluid/operators/collective/c_gen_nccl_id_op.cc:47]: Done

File: paddle/fluid/operators/collective/c_sync_calc_stream_op.cc (hunk 10 above)
Comment (LLMHao) [reply paddle/fluid/operators/collective/c_sync_calc_stream_op.cc:59]: done

File: python/paddle/fluid/framework.py (hunk 12 above)
Comment (LLMHao) [reply python/paddle/fluid/framework.py]: done

File: python/paddle/fluid/transpiler/collective.py [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+                op_role_var = op.all_attrs()[self.op_role_var_key]
+
+                if len(op_role_var) == 0:
+                    continue
+
+                assert len(op_role_var) % 2 == 0
+
+                block._insert_op(
+                    idx + 1,
+                    type='c_sync_calc_stream',
```
Comment (sneaxiy) [inline python/paddle/fluid/transpiler/collective.py:216]: There is no inputs and outputs when appending this op.

File: python/paddle/fluid/transpiler/collective.py [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+                outputs={'Out': var},
+                attrs={
+                    'ring_id': self.global_ring_id,
+                    'root': 0,
+                    self.op_role_key: OpRole.Collective
+                })
+        block.append_op(
+            type='c_sync_comm_stream',
+            attrs={
+                'ring_id': self.global_ring_id,
```
Comment (gavin1332) [inline python/paddle/fluid/transpiler/collective.py:149]: add `inputs={'X': var},outputs={'Out': var}`

File: python/paddle/fluid/transpiler/collective.py (hunk 13 above)
Comment (gavin1332) [reply python/paddle/fluid/transpiler/collective.py:216]: add `inputs={'X': block.vars[grad]},outputs={'Out': block.vars[grad]}`

File: python/paddle/fluid/transpiler/collective.py [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+                        })
+                    offset += 1
+
+        for idx, op in enumerate(block.ops):
+            if self._is_optimizer_op(op):
+                block._insert_op(
+                    idx,
+                    type='c_sync_comm_stream',
+                    attrs={
+                        'ring_id': self.global_ring_id,
```
Comment (gavin1332) [inline python/paddle/fluid/transpiler/collective.py:243]: records the last allreduced grad, and then add it as the input/output of c_sync_comm_stream

File: python/paddle/fluid/transpiler/collective.py [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+                block._insert_op(
+                    idx + 1,
+                    type='elementwise_sub',
+                    inputs={'X': [snapshot],
+                            'Y': [param]},
+                    outputs={'Out': [param]},
+                    attrs={self.op_role_key: OpRole.Collective})
+                block._insert_op(
+                    idx + 2,
+                    type='c_sync_calc_stream',
```
Comment (gavin1332) [inline python/paddle/fluid/transpiler/collective.py:297]: add `inputs={'X': param},outputs={'Out': param}`

File: python/paddle/fluid/transpiler/collective.py [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+                        'reduce_type': 0,
+                        self.op_role_key: OpRole.Collective
+                    })
+
+                ordered_param_snapshot.append((param, snapshot))
+
+        block.append_op(
+            type='c_sync_comm_stream',
+            attrs={
+                'ring_id': self.global_ring_id,
```
Comment (gavin1332) [inline python/paddle/fluid/transpiler/collective.py:318]: record the last allreduced param, and add input/output

**Review by sneaxiy — DISMISSED** [review]
LGTM.

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Pull request merged**